packages/loopring_v3/contracts/test/AmmPool.sol

```diff
@@ -105,6 +146,10 @@ contract AmmPool is IBlockReceiver {
     // A map from a token to the total balance owned directly by LPs (so NOT owned by the pool itself)
```

Comment:
Looks like the locked only used in totalBalance function? My question is that why we divide the add or remove liquidity into two APIs, eg, for add, we have deposit and joinPool. Why not use just only one API like depoistAndJoinPool?

Reply:
`depositAndJoinPool` of course also useful in some cases, so definitely something that should be exposed as well in an actual implementation. But the separation of deposit/withdraw and join/exit is to make things easier and cheaper. Not every join will be successful, and so the funds will already be deposited in the pool contract. If the user wants to join with updated parameters than he only needs to join, not deposit funds again. The same when exiting, funds are still stored in the pool contract (when exiting to L1), so joining again doesn't need (or only partially needs) any new deposited funds.

Join can also be used to just authorize the join, with funds coming from L2 (though it seems I haven't exposed that yet). So in that way it's similar to `approveOffchainTransfer` on the exchange contract so that joins/exits can also be approved by smart contracts. I made join/exit to just be operations that do a basic new thing, similar to how approving a transfer on the exchange doesn't deposit the funds either.